### PR TITLE
Fixes for torch_gather and take_along_axis with broadcast input tensors

### PR DESCRIPTION
**Pull request by naoyam**
There remain some more issues for torch_gather, but the disabled test for take_along_axis is now working.

Haven't tried yet, but hopefully this should also fixes issue #93

### CODE REVIEW

File: csrc/ir_utils.h [hunk 1]
```diff
@@ -364,6 +364,22 @@ TORCH_CUDA_CU_API bool isSqueezeInput(const TensorView* tv);
...
+TORCH_CUDA_CU_API bool isIndexedProducerID(
+    const TensorView* tv,
+    const IterDomain* id);
+
+// Test if the given ID in the given tensor is indirectly written to by,
+// e.g., scatter
+TORCH_CUDA_CU_API bool isIndexedConsumerID(
+    const TensorView* tv,
+    const IterDomain* id);
+
```
Comment (naoyam) [inline csrc/ir_utils.h:382]: Ended up not used but keeping them as they would be handy in some cases

File: csrc/root_domain_map.cpp [hunk 2]
```diff
@@ -1088,11 +1092,21 @@ void ComputeAtRootDomainMapBuilder::handle(TorchGatherOp* op) {
...
+      out_root);
 
-  // Only maps the index root axes. Do not map the input axes due to non-equal
-  // size problem.
-  for (const auto it : c10::irange(idx_root.size())) {
-    setMaybeMapped(idx_td, idx_root[it], out_td, out_root[it]);
+  // Only maps the index root axes unless exact_sizes is true
+  for (const auto i : c10::irange(idx_root.size())) {
+    if (static_cast<int>(i) != op->dim() && op->exactSizes()) {
+      setMaybeMapped(lookup_td, lookup_root[i], out_td, out_root[i]);
```
Comment (naoyam) [inline csrc/root_domain_map.cpp:1107]: This is a bug fix. Was missing before.

File: csrc/root_domain_map.cpp [hunk 3]
```diff
@@ -747,19 +747,20 @@ void ComputeAtRootDomainMapBuilder::initializeBcastMap(
...
+  // pairwise map has no mapping for the broadcast.
+  for (auto consumer : ir_utils::consumerTvsOf(tv)) {
+    const auto p2c =
+        PairwiseRootDomainMap(tv, consumer)
+            .mapProducerToConsumer(tv->domain(), consumer->domain());
+    // Unfortunately, const_cast is required as our const model is
+    // broken.
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+    TORCH_INTERNAL_ASSERT(p2c.find(const_cast<IterDomain*>(id)) == p2c.end());
+  }
```
Comment (naoyam) [inline csrc/root_domain_map.cpp:762]: This is another bug fix. It's just a sanity check of the overall analysis, but this check should be more thorough and should not give false positives.

File: test/test_gather.cpp [hunk 4]
```diff
@@ -489,58 +489,70 @@ TEST_F(IndexingOpTest, TakeAlongBroadcastIndex_CUDA) {
...
-      testValidate(&fusion, cg_outputs, aten_inputs, {ref}, __LINE__, __FILE__);
+TEST_F(IndexingOpTest, GatherBroadcastInput_CUDA) {
+  for (const auto is_take_along : {false, true}) {
+    // torch_gather not supported yet. The issue is one of the index
+    // tensor has a broadcast domain, but its corresponding input
+    // domain is a normal domain. The output domain is also a
+    // broadcast in torch_gather, whereas it's a normal domain in
+    // take_along_axis. In the case of torch_gather, indexing the
+    // input domain needs to be able to index the normal producer
+    // domain with a broadcast reference domain. getProduerIndex needs
```
Comment (zasdfgbnm) [inline test/test_gather.cpp:500]: Just FYI, I would not prioritize fixing this for gather. I don't think the current behavior of gather in torch makes sense. There was some old issue about adding broadcast for gather. https://github.com/pytorch/pytorch/issues/9407

**Review by zasdfgbnm — APPROVED** [review]

**Comment by zasdfgbnm** [comment]
!build

**Pull request merged**